pulp_2to3_migration/app/models/base.py

```diff
@@ -68,93 +72,143 @@ def get_missing_resources(self):
+            ret['repositories_missing_distributors'] = \
+                self.plan_view.repositories_missing_distributors
         return ret
 
 
 class _InternalMigrationPlan:
     def __init__(self, migration_plan):
-        self.migration_plan = migration_plan
-
-        self.plugins_to_migrate = []
-        self.importers_to_migrate = []
-        self.distributors_to_migrate = []
-        # pre-migration *just* needs these repos and nothing else
-        self.repositories_to_migrate = []
+        self._migration_plan = migration_plan
+        self._plugin_plans = []
 
-        # a nested data structure with a format roughly matching the JSON schema.
-        # dictionary where the key is the name of the pulp3 repo and the value is a dict
-        # of other information like repo_versions, importer to use, etc.
-        self.repositories_to_create = {}
+        for plugin_data in self._migration_plan['plugins']:
+            self._plugin_plans.append(PluginMigrationPlan(plugin_data))
 
-        self.missing_importers = []
+        self.repositories_missing_importers = []
         self.missing_repositories = []
-        self.missing_distributors = []
+        self.repositories_missing_distributors = []
 
         # Make sure we've initialized the MongoDB connection first
         connection.initialize()
-        self._populate()
         self._check_missing()
 
-    def _populate(self):
-        for plugin_data in self.migration_plan['plugins']:
-            self.plugins_to_migrate.append(plugin_data['type'])
-            if plugin_data.get('repositories'):
-                self._parse_repository_data(plugin_data.get('repositories'))
-
-    def _parse_repository_data(self, repository_data):
-        for repository in repository_data:
-            name = repository['name']
-
-            _find_importer_repo = repository['pulp2_importer_repository_id']
-            self.importers_to_migrate.append(_find_importer_repo)
-
-            repository_versions = self._parse_repository_version_data(
-                repository.get('repository_versions', [])
-            )
-
-            repository_data = {
-                "pulp2_importer_repository_id": _find_importer_repo,
-                "versions": repository_versions
-            }
-
-            self.repositories_to_create[name] = repository_data
-
-    def _parse_repository_version_data(self, repository_version_data):
-        repository_versions = []
-
-        for repository_version in repository_version_data:
-            pulp2_repository_id = repository_version['pulp2_repository_id']
-            self.repositories_to_migrate.append(pulp2_repository_id)
-            repository_versions.append(pulp2_repository_id)
+    @property
+    def all_repositories_importers_to_migrate(self):
+        # flat list of all importers to migrate
+        return itertools.chain.from_iterable(
+            [plugin.repositories_importers_to_migrate for plugin in self._plugin_plans]
```

Comment:
If it happens that plan doesn't provide with any data and we chain empty list, there is still iterable and later we can't check if it's empty or not.
It creates a problem down the line when we conditionally create a mongo query.

Reply:
Do you think wrapping this with a call to ```list()``` would fix that?

Reply:
yes, it should do